src/storage/src/render/persist_sink.rs

```diff
@@ -7,113 +7,824 @@
+                        }
+                    }
+                }
+                else => {
+                    // All inputs are exhausted, so we can shut down.
+                    return;
+                }
+            }
+
+            // We may have the opportunity to commit updates.
+            if PartialOrder::less_equal(&current_upper, &desired_frontier) {
+                trace!(
+                    "persist_sink {collection_id}/{shard_id}: \
+                        CAN emit: \
+                        current_upper: {:?}, \
+                        desired_frontier: {:?}",
+                    current_upper,
+                    desired_frontier
+                );
+
+                trace!(
+                    "persist_sink {collection_id}/{shard_id}: \
+                        in-flight batches: {:?}, \
+                        batch_descriptions_frontier: {:?}, \
+                        desired_frontier: {:?}",
+                    in_flight_batches,
+                    batch_descriptions_frontier,
+                    desired_frontier,
+                );
+
+                // We can write updates for a given batch description when
+                // a) the batch is not beyond `batch_descriptions_frontier`,
+                // and b) we know that we have seen all updates that would
+                // fall into the batch, from `desired_frontier`.
+                let ready_batches = in_flight_batches
+                    .keys()
+                    .filter(|(lower, upper)| {
+                        !PartialOrder::less_equal(&batch_descriptions_frontier, lower)
+                            && !PartialOrder::less_than(&desired_frontier, upper)
+                    })
+                    .cloned()
+                    .collect::<Vec<_>>();
+
+                trace!(
+                    "persist_sink {collection_id}/{shard_id}: \
+                        ready batches: {:?}",
+                    ready_batches,
+                );
+
+                for batch_description in ready_batches.into_iter() {
+                    let cap = in_flight_batches.remove(&batch_description).unwrap();
+
+                    if collection_id.is_user() {
+                        trace!(
+                            "persist_sink {collection_id}/{shard_id}: \
+                                emitting done batch: {:?}, cap: {:?}",
+                            batch_description,
+                            cap
+                        );
+                    }
+
+                    let (batch_lower, batch_upper) = batch_description;
+
+                    let mut finalized_timestamps: Vec<_> = stashed_batches
+                        .keys()
+                        .filter(|time| {
+                            batch_lower.less_equal(time) && !batch_upper.less_equal(time)
+                        })
+                        .copied()
+                        .collect();
+
+                    // TODO(guswynn): is this sorting required?
```

Comment:
don't think so, am I missing something?

Reply:
I think in the past it was...

Reply:
maybe it's the latter then. do you remember why?